vast/vast.cpp

```diff
@@ -37,89 +37,110 @@
 #include <string>
 #include <vector>
 
-using namespace vast;
-using namespace vast::system;
+namespace vast {
+
+caf::expected<std::vector<std::filesystem::path>>
+load_plugins(system::configuration& cfg) {
+  auto loaded_plugin_paths = std::vector<std::filesystem::path>{};
+  auto& plugins = plugins::get();
+  // Load plugins.
+  auto plugin_paths_or_names
+    = caf::get_or(cfg, "vast.plugins", std::vector<std::string>{});
+  const auto check_and_remove_plugin = [&](const char* name) {
+    auto bundled = std::remove(plugin_paths_or_names.begin(),
+                               plugin_paths_or_names.end(), name);
+    const auto result = bundled != plugin_paths_or_names.end();
+    plugin_paths_or_names.erase(bundled, plugin_paths_or_names.end());
+    return result;
+  };
+  if (check_and_remove_plugin("all")) {
+    // The special plugin name "bundled" enables all plugins, i.e., plugins
+    // built alongside VAST *and* external dynamic plugins.
+  } else {
+    // The special plugin name "bundled" enables all bundled plugins, i.e.,
+    // plugins built alongside VAST.
+    if (check_and_remove_plugin("bundled"))
+      for (auto&& plugin_name : std::vector<std::string>{VAST_BUNDLED_PLUGINS})
+        if (std::none_of(plugin_paths_or_names.begin(),
+                         plugin_paths_or_names.end(), [&](auto&& x) {
+                           return x == plugin_name;
+                         }))
+          plugin_paths_or_names.push_back(std::move(plugin_name));
+    // Check if any of the specified plugins is already loaded and remove it
+    // from the list of specified plugins if that's the case. If bare mode is
+    // enabled, unload plugins that were not specified explicitly to avoid
+    // accidentaly running with undesired plugins.
+    for (auto& plugin : plugins) {
+      VAST_ASSERT(plugin.type() != plugin_ptr::type::dynamic,
+                  "dynamic plugins must be loaded at this point");
+      if (auto it = std::find_if(plugin_paths_or_names.begin(),
+                                 plugin_paths_or_names.end(),
+                                 [&](const auto& plugin_path_or_name) {
+                                   return plugin->name() == plugin_path_or_name;
+                                 });
+          it != plugin_paths_or_names.end())
+        plugin_paths_or_names.erase(it);
+      else if (plugin.type() != plugin_ptr::type::native)
+        plugin = {};
+    }
+    // Remove all possibly invalidated static plugins.
+    plugins.erase(std::remove_if(plugins.begin(), plugins.end(),
+                                 [](const auto& plugin) {
+                                   return !plugin;
+                                 }),
+                  plugins.end());
+    // Load the specified dynamic plugins.
+    for (const auto& plugin_path_or_name : plugin_paths_or_names) {
+      if (auto loaded_plugin = detail::load_plugin(plugin_path_or_name, cfg)) {
```

Comment:
I'm confused, isn't this still in the "else" part, i.e. only hit when we did not specify "all" as plugin name?